Lombiq.Hosting.Tenants.FeaturesGuard/Handlers/FeaturesEventHandler.cs

```diff
@@ -30,53 +38,193 @@ public sealed class FeaturesEventHandler : IFeatureEventHandler
+    Task IFeatureEventHandler.EnabledAsync(IFeatureInfo feature) => EnableConditionallyEnabledFeaturesAsync(feature);
 
     Task IFeatureEventHandler.DisablingAsync(IFeatureInfo feature) => Task.CompletedTask;
 
-    Task IFeatureEventHandler.DisabledAsync(IFeatureInfo feature) => KeepFeaturesEnabledAsync(feature);
+    async Task IFeatureEventHandler.DisabledAsync(IFeatureInfo feature)
+    {
+        await KeepConditionallyEnabledFeaturesEnabledAsync(feature);
+        await DisableConditionallyEnabledFeaturesAsync(feature);
+    }
 
     Task IFeatureEventHandler.UninstallingAsync(IFeatureInfo feature) => Task.CompletedTask;
 
     Task IFeatureEventHandler.UninstalledAsync(IFeatureInfo feature) => Task.CompletedTask;
 
-    public async Task EnableMediaRelatedFeaturesAsync(IFeatureInfo featureInfo)
+    /// <summary>
+    /// Enables conditional features (key) if one of their corresponding condition features (value) was enabled.
+    /// </summary>
+    /// <param name="featureInfo">The feature that was just enabled.</param>
+    public async Task EnableConditionallyEnabledFeaturesAsync(IFeatureInfo featureInfo)
     {
-        if (featureInfo.Id is not FeatureNames.Media and not FeatureNames.MediaCache) return;
+        if (_shellSettings.IsDefaultShell() ||
+            _conditionallyEnabledFeaturesOptions.Value.EnableFeatureIfOtherFeatureIsEnabled is not { } conditionallyEnabledFeatures)
+        {
+            return;
+        }
 
+        var allConditionFeatureIds = new List<string>();
+        foreach (var conditionFeatureId in conditionallyEnabledFeatures.Values)
+        {
+            var separatedConditionFeatureIds = conditionFeatureId.SplitByCommas();
+            allConditionFeatureIds.AddRange(separatedConditionFeatureIds.Select(id => id.Trim()));
+        }
+
+        if (!allConditionFeatureIds.Contains(featureInfo.Id))
+        {
+            return;
+        }
+
+        // Enable conditional features if they are not already enabled.
         var allFeatures = await _shellFeaturesManager.GetAvailableFeaturesAsync();
 
-        if (featureInfo.Id == FeatureNames.Media)
+        var conditionalFeatureIds = new List<string>();
+        conditionallyEnabledFeatures.ForEach(keyValuePair =>
+        {
+            var valueFormatted = new List<string>();
+            var separatedValues = keyValuePair.Value.SplitByCommas();
+            foreach (var separatedValue in separatedValues)
+            {
+                valueFormatted.Add(separatedValue.Trim());
+            }
+
+            if (valueFormatted.Contains(featureInfo.Id))
+            {
+                conditionalFeatureIds.Add(keyValuePair.Key);
+            }
+        });
+
+        var conditionalFeatures = allFeatures.Where(feature => conditionalFeatureIds.Contains(feature.Id));
+        var currentlyDisabledFeatures = await _shellFeaturesManager.GetDisabledFeaturesAsync();
```

Comment:
What happens if you just enable all the features that need to be kept enabled without checking if they need to be enabled first? Isn't Orchard doing that already? Similarly in `DisableConditionallyEnabledFeaturesAsync`.

Reply:
Well, the universe has not yet imploded despite my best efforts while testing it without these checks. It looks like they can be removed indeed.